docs/developers/architecture/magicgui_type_reg.md

```diff
@@ -0,0 +1,71 @@
+(magicgui_type_registration)=
+
+# `magicgui` type registration
+
+[`magicgui`](https://pyapp-kit.github.io/magicgui/) uses
+[type hints](https://peps.python.org/pep-0484/) to infer the appropriate widget type
+for a given function parameter. It allows third party packages
+(like `napari`) to [register](https://pyapp-kit.github.io/magicgui/type_map/#registering-support-for-custom-types) support for their types using
+{func}`~magicgui.type_map.register_type`. `napari` registers
+a number of types, additionally specifying, where appropriate:
+
+* widget type for the parameter type
+* function for updating inputs for the widget
+* return callback, for return types
+
+This enables `magicgui` widgets to be easily created via type annotations.
+
+```{note}
+This page provides implementation details on `napari`-specific type registration
+in `magicgui` and is aimed at `napari` developers.
+
+For information about using `magicgui` (for users and plugin developers) see
+[](creating-widgets).
+```
+
+## Registration details
```

Comment:
Would be good to have some subheadings here maybe? One for `layers` and one for `viewer` at least?